### PR TITLE
Created SegmentTokenizer

### PR DESCRIPTION
**Pull request by avarun42**

### CODE REVIEW

File: QueryBuilder/SegmentTokenizer.js [hunk 1]
```diff
@@ -0,0 +1,85 @@
+/**
+ * SegmentTokenizer.js
+ */
+
+'use strict';
+
+const SEGMENT_REGEXP = /^(!?)([a-z\-]{1,10})([\d]{1,6})([a-z]{0,4})(\+?)$/i;
```
Comment (brenmcnamara) [inline QueryBuilder/SegmentTokenizer.js]: There's gonna be a lot of other types of queries besides "+".

I'd recommend separating this out:

`QUERY_REGEXP = /[\+\*]$/`

and you can concatenate regexp like this:

`var concatenatedRegexp = new RegExp(regexp1.source + regexp2.source);`

File: QueryBuilder/SegmentTokenizer.js [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+   * @return {Object} The tokenized segment.
+   */
+  tokenize(segment) {
+    let match = segment.match(SEGMENT_REGEXP);
+    if (!match) {
+      throw Error(`Invalid segment ${segment}`);
+    }
+
+    let majorCode = match[2].toUpperCase();
+    let courseNumber = match[3];
```
Comment (brenmcnamara) [inline QueryBuilder/SegmentTokenizer.js]: cast this to a number

File: QueryBuilder/SegmentTokenizer.js [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+
+    let returnObject = {
+      majorCode,
+      courseNumber,
+      courseSuffix,
+      queryType,
+      negate
+    };
+
+    return returnObject;
```
Comment (brenmcnamara) [inline QueryBuilder/SegmentTokenizer.js]: you can shorten this with:

return {
  majorCode,
  courseNumber,
  courseSuffix,
  queryType,
  negate,
};

File: QueryBuilder/SegmentTokenizer.js [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+    let match = segment.match(SEGMENT_REGEXP);
+    if (!match) {
+      throw Error(`Invalid segment ${segment}`);
+    }
+
+    let majorCode = match[2].toUpperCase();
+    let courseNumber = match[3];
+    let courseSuffix = match[4].length ? match[4].toUpperCase() : null;
+    let queryType = match[5].length ? match[5] : null;
+    let negate = match[1].length ? true : false;
```
Comment (brenmcnamara) [inline QueryBuilder/SegmentTokenizer.js]: `let negate = match[1].length > 0;`

File: QueryBuilder/SegmentTokenizer.js (hunk 3 above)
Comment (avarun42) [inline QueryBuilder/SegmentTokenizer.js]: Funnily enough, that's what I had before, then I changed it because I thought this way was more flexible.

**Pull request merged**